feat: Start react testing library guide (#458)
getsentry/develop: src/docs/frontend/using-rtl.mdx

---
title: Using React Testing Library
---

We are in the process of converting our tests from Enzyme to React Testing Library.

> In this guide, you'll find tips to follow best practices and anticipate common pitfalls.

Reviewer: ```suggestion
In this guide, you'll find tips to follow best practices and avoid common pitfalls.
```

We have two ESLint rules in place to help with this:

- [eslint-plugin-jest-dom](https://github.com/testing-library/eslint-plugin-jest-dom)
- [eslint-plugin-testing-library](https://github.com/testing-library/eslint-plugin-testing-library)

> We strive to write tests in such a way that closely resembles how our application is used.

Reviewer: ```suggestion
We strive to write tests in a way that closely resembles how our application is used.
```

> So rather than dealing with instances of rendered components, we query the DOM in the same way the user would.

Reviewer: ```suggestion
Rather than dealing with instances of rendered components, we query the DOM in the same way the user would.
```

We find form elements by their label text (just like a user would), we find links and buttons from their text (like a user would).

As a part of this goal, we avoid testing implementation details so refactors (changes to implementation but not functionality) don't break the tests.

We are generally in favor of Use Case Coverage over Code Coverage.

## Querying

- use `getBy...` as much as possible
- use `queryBy...` only when checking for non-existence

> - use `await findBy...` only when expecting an element to appear after but the change to the DOM might not happen immediately

Reviewer: ```suggestion
- use `await findBy...` only when expecting an element to appear after a change to the DOM that might not happen immediately
```

The `getBy...` and `findBy...` variants will throw a more helpful error message if no element is found.

```javascript
import { mountWithTheme, screen } from "sentry-test/reactTestingLibrary";

// ❌
mountWithTheme(<Example />);
expect(screen.queryByRole("alert")).toBeInTheDocument();

// ✅
mountWithTheme(<Example />);
expect(screen.getByRole("alert")).toBeInTheDocument();
expect(screen.queryByRole("button")).not.toBeInTheDocument();
```

Avoid using `waitFor` for waiting on appearance, use `findBy...` instead ([examples](https://github.com/getsentry/sentry/pull/29544)).
These two are basically equivalent, but the `findBy...` is simpler and the error message we get will be better.

```javascript
import {
  mountWithTheme,
  screen,
  waitFor,
} from "sentry-test/reactTestingLibrary";

// ❌
mountWithTheme(<Example />);
await waitFor(() => {
  expect(screen.getByRole("alert")).toBeInTheDocument();
});

// ✅
mountWithTheme(<Example />);
expect(await screen.findByRole("alert")).toBeInTheDocument();
```

Avoid using `waitFor` for waiting on disappearance, use `waitForElementToBeRemoved` instead ([examples](https://github.com/getsentry/sentry/pull/29547)).

> The latter is using MutationObserver which is more efficient than polling the DOM at regular intervals with waitFor.

Reviewer: ```suggestion
The latter is uses `MutationObserver` which is more efficient than polling the DOM at regular intervals with waitFor.
```
